### DISCUSSION TITLE
What is abstraction?

### DISCUSSION THREAD

**Discussion by mishelshaji** [mishelshaji/Learn-DotNet#12] [Q&A] [answered]
Explain the consept of abstraction in OOP with examples.

**Comment by thanoofnaina** [comment]
### Abstraction
Abstraction is showing only the needful things and hiding every other internal things or working of the code. Like showing the button in an application or a web page and hiding the process behind the clicking of button to a user.
 
### Example
For using an Exception class, one doesn't need to know about what happens inside the exception class.

```csharp
catch (Exception ex)
{
    Console.WriteLine(ex.Message);
}
```

**Comment by kavitha-BS** [comment]
Abstraction is the process of hiding the internal details of an application from the outer world. Abstraction is used to describe things in simple terms. It’s used to create a boundary between the application and the client programs.
eg:
abstract class Animal
{
 
  public abstract void animalSound();
 
  public void sleep()
  {
    Console.WriteLine("Zzz");
  }
}

// Derived class (inherit from Animal)
class Pig : Animal
{
  public override void animalSound()
  {
   
    Console.WriteLine("The pig says: wee wee");
  }
}

class Program
{
  static void Main(string[] args)
  {
    Pig myPig = new Pig(); 
    myPig.animalSound();  // Call the abstract method
    myPig.sleep();  // Call the regular method
  }
}

**Comment by Nandana-04** [comment]
# Abstraction

Abstraction is the process of hiding certain details and showing only essential information to the user. 
 The abstract keyword is used for classes and methods:

-   Abstract class: is a restricted class that cannot be used to create objects (to access it, it must be inherited from another class).
-   Abstract method: can only be used in an abstract class, and it does not have a body. The body is provided by the derived class (inherited from).

eg :
```csharp
// Abstract class
   abstract class Fruit
      {
       //  Abstract method (does not have a body)
         public abstract void fruitColor();
       // Regular Method
          public void place()
             {
                  Console.WriteLine("Available in Jammu Kashmir");
             }
      }
// Derived class (inherit from Fruit)
    class Apple : Fruit
    {
        public override void fruitColor()
        {
            Console.WriteLine("Apple is Red , Green , Yellow .");
        }

    }

Apple choice = new Apple(); // Create an object .
choice.fruitColor(); // Call the abstract method .
choice.place(); // Call the regular method .

```

**Reply by mishelshaji** [reply]
The question was about abstraction. Abstract class is a different thing.

**Comment by ARAVIND0407** [comment]
# Abstraction
Abstraction is the process of hiding the internal details of an application from the outer world. Abstraction is used to describe things in simple terms. It’s used to create a boundary between the application and the client programs.

An object contains some properties and methods. We can hide them from the outer world through access modifiers. We can provide access only for required functions and properties to the other programs. This is the general procedure to implement abstraction in OOPS.

**Comment by rashid234** [comment]
# Abstraction
Abstraction in C#is the process to hide the internal details and show only the functionality. The abstract modifier indicates the incomplete implementation. The keyword abstract is used before the class or method to declare the class or method as abstract. Also, the abstract modifier can be used with indexers, events, and properties. 

Example: 

public abstract void grett();
// this indicates the method 'grett()' is abstract
 
abstract class gfg
// this indicates the class 'gfg' is abstract

**Comment by riya45810** [comment]
### C# Abstraction
Data abstraction is the process of hiding certain unwanted details and only shows essential information to the user.it can be done with either **abstract classes** or **interfaces**.the **abstract** keyword is used for classes and methods.
**Abstract class:**  is a restricted class that cannot be used to create objects, to access it must inherited from another class.
**Abstact method:**  can only be used in an abstract class, and it does not have any body. The body provided by the derived class.
```csharp
abstract class Car{
public abstract void carBrand();
public void color(){
 Console.WriteLine("Black");
}
}
class Benz : Car{
public override void carBrand(){
Console.WriteLine("Mercedes-Benz");}
}
class Program
{
public void Main(string[] args)
{
Benz bz = new Benz();
bz.carBrand();
bz.color();
}
}
```

**Comment by sreelakshmiChandran** [comment]
# Abstraction

Abstraction is one of the principles of the oop concept. Abstraction is the principle that hides the inner activities of a program or process from the outside world. It shows only the essential information to the user.

To achieve security - hide certain details and only show the important details of an object.

```csharp
abstract class Dog
{
   public abstract void breed();
   public void color()
   {
       Console.WriteLine("Black");
   }
}
    class Labrador : Dog 
{
    public override void breed(){
    Console.WriteLine("Labrador");}
}
class Program
{
    public void Main(string[] args)
   {
      Labrador s = new Labrador();
      s.breed();
      s.color();
   }
}
```

**Comment by akhilamp123** [comment]
### Abstraction
Abstraction in Object-Oriented Programming (OOP) is the process of hiding the internal details of an object and showing only the necessary information to the user. It allows the user to interact with the object without having to know the specific implementation details.

```csharp
abstract class Animal
    {
        public abstract void MakeSound();

        public void Run()
        {
            Console.WriteLine("Animals can run");
        }

        public class Cat: Animal
        {
            public override void MakeSound()
            {
                Console.WriteLine("Meow Meow");
            }
        }
    }
Animal animal = new Cat();
animal.MakeSound();
animal.Run();

````

**Comment by sarath534** [comment]
# Abstraction in c#
In C#, data abstraction is the process of hiding the implementation details of an object and exposing only the necessary information to the user. This is achieved through the use of interfaces and abstract classes.

For example, consider a class "Car" that has methods such as "startEngine()" and "stopEngine()" and properties such as "engineRunning" and "currentSpeed". The user only needs to know that they can start and stop the engine and check the status of the engine and speed, but they don't need to know the specific implementation details of how the engine starts or stops.

An abstract class can be used to define the common interface for different types of cars and hide the implementation details.
```csharp
abstract class Car
{
    public abstract void startEngine();
    public abstract void stopEngine();
    public bool engineRunning {get; protected set;}
    public int currentSpeed {get; protected set;}
}
class ElectricCar : Car
{
    public override void startEngine()
    {
        // specific implementation for starting an electric car engine
    }
    public override void stopEngine()
    {
        // specific implementation for stopping an electric car engine
    }
}
```

**Comment by Jiby-Joseph-123** [comment, marked as answer]
# Abstraction

Data abstraction Is the process of hiding certain details and showing only essential information to the user. 

Abstract class: It is a restricted class that cannot be used to create objects.

Abstract method: can only be used in an abstract class, and it does not have a body. The derived class provides the body.

An abstract class can have both abstract and regular methods:

To access the abstract class, it must be inherited from another class. To inherit a class we use the ":" symbol, and that override keyword to override the base class method.
```csharp
abstract class Animal // abstract class.
{
    public abstract void animalSound(); // abstract method.

    public void sleep() //regular method.
    {
        Console.Write("Zzz");
    }
}

class Pig : Animal
{
    public override void animalSound()
    {
        Console.Write("Sound of pig!");
    }
}
class MainClass
{
    static void Main()
    {
        Pig pig = new Pig();
        pig.animalSound();
        pig.sleep();
    }
}
```

**Comment by AnitaMaryThomas** [comment]
### **#Data Abstraction**

Data Abstraction is the process of hiding certain information and only showing essential information to the user.
Abstraction can be achieved with either abstract class or interfaces.

#**Abstract Class**
 A restricted class tha cannot be used to create objects.
#**Abstract Method**
It only be used in an abstract class, and it kdoes not have a body. The body is provided by the derived class.

**#Example:**
```csharp
class Program 
{
abstract class Animal
{
public abstract void Eat();
public void Sound()
{
Console.WriteLine("Animal can sound");
}
}
class Lion : Animal
{
public override void Eat()
{
Console.WriteLine("Lion can eat");
}
}
static void Main(string[] args)
{
Lion lion = new Lion();
Animal animal = lion;
animal.Eat();
lion.Sound();
}
}
```

**Comment by SreelakshmiSajeev** [comment]
#Abstraction

Abstraction is the process of hiding certain details and showing only the essential details to the users. It can be achieved either through abstract classes or interfaces.

#Abstract Classes : These types of classes can be used accessed only  by inheriting it from another class. It cannot be used to create objects.

#Abstract Methods : These types of method contains only declaration .It's definition is provided in the derived class.

#Example.

```C#
using System;

namespace MyApplication
{
  abstract class Animal
  {
    public abstract void animalDance();
  
  }
  class Dog : Animal
  {
    public override void animalDance()
    {
      Console.WriteLine("Hi I can dance...........");
    }
  }

  class Program
  {
    static void Main(string[] args)
    {
      Dog dog = new Dog();  
      dog.animalDance();
  
    }
  }
}

**Comment by amalapaul** [comment]
## ABSTRACTION

- Abstraction is the act of removing characteristics from something in order to reduce it to a set of essential functions or characteristics.

- In OOP abstraction concept, the actual implementation is hidden from the user and only required functionality will be available to the user.
 
- It is a powerful methodology to manage complex systems. Abstraction is managed by well-defined objects and their hierarchical classification.
 
- The concept is mainly used in the project design level, here in program design, what features a program will contain is declared but not how to implement the features.



Example of abstraction is a watch, you checks the time many times in a day but you never need to know what the internal mechanism of the watch is.

```csharp
interface Car
{
public abstract int topSpeed();
public abstract float mileage();
}
```

In the above example we can see what function an implementing class will contain but we cannot see how the function is implemented.

**Comment by ManikuttyBabu** [comment]
- Abstraction is the process which hiding unreliable data or information who are not related to use but is still present over there.
- Abstract Classes: These types of classes can be used and accessed only by inheriting them from another class. It cannot be used to create objects.

- Abstract Methods: These types of method contains only declaration. Its definition is provided in the derived class.
 
#Example. 
```C#
using System;
 namespace Myapp;
{
  abstract class Vehicles
  {
    public abstract void vehicleColour();
  }
  class Car  : Vehicle
  {
    public override void  vehicleColour()
    {
      Console.WriteLine("The vehicle is in red colour");
    }
  }  
 class Program
  {
    static void Main(string[] args)
    {
     Car car = new Car();  
      car.vehicleColour();
    }
  }
}

**Comment by AnitaMaryThomas** [comment]
### **Data Abstraction**

- Data Abstraction is the process of hiding certain information and only showing essential information to the user.
- Abstraction can be achieved with either abstract class or interfaces.

 **Abstract Class**

- A restricted class that cannot be used to create objects.

 **Abstract Method**

- It only be used in an abstract class, and it does not have a body. The body is provided by the derived class.

**Example:**

```csharp
class Program 
{
abstract class Animal
{
public abstract void Eat();
public void Sound()
{
Console.WriteLine("Animal can sound");
}
}
class Lion : Animal
{
public override void Eat()
{
Console.WriteLine("Lion can eat");
}
}
static void Main(string[] args)
{
Lion lion = new Lion();
Animal animal = lion;
animal.Eat();
lion.Sound();
}
}
```
